Add unit tests for helpermock.call.RequestMock

ThomasPerkins1123 opened the pull request:
Hi,

I've analysed your code base and noticed that:
`br.com.conductor.heimdall.middleware.util.helpermock.call.RequestMock` 
in the heimdall-middleware-spec module is not fully tested.

I've written some tests that cover this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests should help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other particular classes that you consider important.

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,86 @@
+    final Query retval = objectUnderTest.query();
+
+    Assert.assertNotNull(retval);
+    final HashMap<String, String> hashMap = new HashMap<String, String>();
```
Since its an empty hashmap you could do
`Assert.assertEquals(new HashMap<String, String>(), retval.getAll());`

and get rid of this line` final HashMap<String, String> hashMap = new HashMap<String, String>();`

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,86 @@
+
+  @Test
+  public void queryOutputNotNull() {
+
```
and remove the extra white spaces to make it look shorter.
add some between different asserts :)

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,86 @@
+    final RequestMock objectUnderTest = new RequestMock();
+    Assert.assertEquals(null, objectUnderTest.pathParam(""));
+  }
+}
```
add a extra line at the bottom :)

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.HashMap;
+
+public class RequestMockTest {
+
```
You could import a static assert and the test class to reduce the length of the line. They try to keep the length less than 80 characters
`import static org.junit.Assert.*;`

so your line would be `assertEquals()` rather than `Assert.assertEquals()`

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.Timeout;
```
Line 8 (Timeout) can be removed.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+public class RequestMockTest {
+
+  @Rule public final ExpectedException thrown = ExpectedException.none();
+  @Rule public final Timeout globalTimeout = new Timeout(10000);
```
Line 16 (Timeout) can be removed.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+
+public class RequestMockTest {
+
+  @Rule public final ExpectedException thrown = ExpectedException.none();
```
Line 15 is not needed if no exceptions are thrown.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+import br.com.conductor.heimdall.middleware.spec.Query;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
```
Line 7 is not needed if no exceptions are thrown.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+
+import br.com.conductor.heimdall.middleware.spec.Header;
+import br.com.conductor.heimdall.middleware.spec.Query;
+import org.junit.Rule;
```
Line 5 can be removed if the @Rule lines later are removed.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+  @Test
+  public void testHeaderOutputNotNull() {
+    final RequestMock objectUnderTest = new RequestMock();
+
```
Remove blank lines except before Asserts (see style guide)

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+
+    final Header retval = objectUnderTest.header();
+
+    assertNotNull(retval);
```
Don't use assertNotNull() if there is another assert that effectively checks the same thing.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+    final Header retval = objectUnderTest.header();
+
+    assertNotNull(retval);
+    assertEquals(new HashMap<String, String>(), retval.getAll());
```
Inline variables if possible (objectUnderTest() and retval) while keeping lines down to 80 chars.
Avoid the name "retval".

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+    final RequestMock objectUnderTest = new RequestMock();
+    objectUnderTest.setSendResponse(false);
+
+    // Method returns void, testing that no exception is thrown
```
We don't usually test methods which don't return a value.

peterschrammel replied:
This should be tested, but you have to check for an observable side effect.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+    assertEquals("testUrl", objectUnderTest.getUrl());
+  }
+
+  @Test
```
Can we get another test or two for setPathParam()?

petersear reviewed (changes requested):
I've added some comments to the first test which apply to some of the others (delete blank lines, inline values where possible, see style guide)
In general, we don't test simple setters or getters as they are not very interesting.

petersear commented:
You will want to squash your final commits to one.
"Diffblue Cover" in the PR text should link to the website.
Add class name to PR title.

peterschrammel commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+  public void testSetUrl() {
+    final RequestMock objectUnderTest = new RequestMock();
+
+    objectUnderTest.setUrl("testUrl");
```
Use a proper url

peterschrammel commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,79 @@
+  @Test
+  public void testSetPathParam() {
+    final RequestMock objectUnderTest = new RequestMock();
+    assertEquals(null, objectUnderTest.pathParam(""));
```
Maybe also with a proper param?

ThomasPerkins1123 commented:
Changes made please re-review

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+
+import org.junit.Test;
+import static org.junit.Assert.*;
+
```
remove space

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+  @Test
+  public void testHeaderOutputNotNull() {
+    final RequestMock requestMock = new RequestMock();
+
```
remove space

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+public class RequestMockTest {
+
+  @Test
+  public void testHeaderOutputNotNull() {
```
change the test case name to testHeader. So the test case name is actually the method you are testing

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+  @Test
+  public void testQueryOutputNotNull() {
+    final RequestMock requestMock = new RequestMock();
+
```
remove space

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  @Test
+  public void testQueryOutputNotNull() {
```
same as before. test case name

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+  @Test
+  public void testSetAppNameInput() {
+    final RequestMock requestMock = new RequestMock();
+
```
remove space

Braavos6 commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+    final RequestMock requestMock = new RequestMock();
+    assertEquals(null, requestMock.pathParam("1643/kajd"));
+  }
+}
```
add space at the bottom

Braavos6 reviewed (changes requested):
Few changes :)

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+package br.com.conductor.heimdall.middleware.util.helpermock.call;
+
+import org.junit.Test;
+import static org.junit.Assert.*;
```
Put import static line(s) in a separate section below the other imports.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+  @Test
+  public void testSetBodyInput() {
+    final RequestMock requestMock = new RequestMock();
+
```
Remove blank line

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+  @Test
+  public void testSetUrl() {
+    final RequestMock requestMock = new RequestMock();
+
```
Remove blank line

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  @Test
+  public void testSetPathParam2() {
```
Combine the two testSetPathParam() tests into one

petersear reviewed (changes requested):
See style guide for space around assertions:
https://github.com/diffblue/customer-experience/blob/master/diffblue-yml-pr/checklist.md#assertions

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,50 @@
+import org.junit.Test;
+import java.util.HashMap;
+
+import static org.junit.Assert.*;
```
I think this static import is probably ok. I've just imported each Assert method separately in a recent PR to save importing unnecessary stuff but, looking at other tests in this project, I can see examples of both as well another case where org.junit/Assert is imported and each assert is prefixed "Assert."

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  @Test
+  public void testSetAppNameInput() {
```
I'd remove "Input" from the test name.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  @Test
+  public void testSetBodyInput() {
```
Remove "Input" from test name.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  @Test
+  public void testSetPathParam() {
```
The method under test is pathParam() so the test name should be testPathParam()

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,50 @@
+  @Test
+  public void testSetPathParam() {
+    final RequestMock requestMock = new RequestMock();
+    assertEquals(null, requestMock.pathParam(""));
```
Apologies, I think I missed this earlier: use assertNull(foo) instead of assertEquals(null, foo)
(similarly with assertTrue and assertFalse() if the situation arises)

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,50 @@
+  public void testSetPathParam() {
+    final RequestMock requestMock = new RequestMock();
+    assertEquals(null, requestMock.pathParam(""));
+    assertEquals(null, requestMock.pathParam("testPath"));
```
Do these 3 tests cover all the lines in the method?
If so, are all 3 needed for that purpose?

ThomasPerkins1123 replied:
No the 3 tests do not cover any extra lines but they where added as you suggested adding more tests in a previous comment.
A single test would cover most of this function but the final part of the function is unreachable without reflection.

petersear replied:
OK I should have been more clear that by asking if more tests could be added, I meant tests which would add coverage as opposed to more tests to cover the same lines. If no more can be added without using reflection then I would keep with the minimum number.

petersear commented:
btw we were talking about the PR title, which is quite long with the fully qualified class name. I found this article:
[https://chris.beams.io/posts/git-commit](https://chris.beams.io/posts/git-commit)
which suggests limiting the first line of a commit to 50 characters. That could then be applied to the PR title, in which case maybe include the unqualified class name in both (the FQ name is in the PR text). Also, I would suggest surrounding the class name with back ticks (`) to make it more readable.

ThomasPerkins1123 commented:
Okay 50 charters seems reasonable but is quite small so what should this name be refactored to? 
would something like this "Add unit tests for RequestMock in middleware.util"  be appropriate or should i try to take more from the path such as "Add unit tests for br.com.conductor.heimdall.middleware"

JohnLBergqvist commented:
@ThomasPerkins1123 I would go with `Add unit tests for helpermock.call.RequestMock` for the 1st line of the commit message.

JohnLBergqvist commented on heimdall-middleware-spec/pom.xml:
```diff
@@ -60,4 +60,4 @@
 	<description>This module is a specification of a Middleware structure.</description>
-</project>
\ No newline at end of file
+</project>
```
add a blank line on the end of this - this should then remove the file from the list of changes.

petersear commented:
Agree with John's suggestion on the first line of the commit message.

petersear commented on heimdall-middleware-spec/src/test/java/br/com/conductor/heimdall/middleware/util/helpermock/call/RequestMockTest.java:
```diff
@@ -0,0 +1,52 @@
+  @Test
+  public void testPathParam() {
+    final RequestMock requestMock = new RequestMock();
+    assertNull(null, requestMock.pathParam(""));
```
I think these assertNull()s were originally assertEquals()
Having made the change, you can drop the first argument (null)

ThomasPerkins1123 committed:
Add unit tests for helpermock.call.RequestMock

Adds unit tests for:
br.com.conductor.heimdall.middleware.util.helpermock.call.RequestMock

These tests were written using Diffblue Cover.
